pmlc/dialect/tile/transforms/expand_reshape.cc

```diff
@@ -0,0 +1,233 @@
+// Copyright 2020, Intel Corporation
+
+#include "mlir/Pass/Pass.h"
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "pmlc/dialect/stdx/ir/ops.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+
+#include <numeric>
+
+using namespace mlir;                // NOLINT
+using namespace pmlc::dialect::stdx; // NOLINT
+
+namespace pmlc::dialect::tile {
+
+// Compute the number of indices needed
+int computeNumIdxs(ArrayRef<int64_t> srcShape, ArrayRef<int64_t> dstShape) {
+  SmallVector<int64_t, 4> srcDims(srcShape.begin(), srcShape.end());
+  SmallVector<int64_t, 4> dstDims(dstShape.begin(), dstShape.end());
+  int pSrc = srcDims.size() - 1;
+  int pDst = dstDims.size() - 1;
+  int numIdxs = 0;
+  while (pSrc >= 0 && pDst >= 0) {
+    while (pSrc >= 0 && srcDims[pSrc] == 1) {
+      --pSrc;
+    }
+    if (pSrc < 0) {
+      break;
+    }
+    while (pDst >= 0 && dstDims[pDst] == 1) {
+      --pDst;
+    }
+    if (pDst < 0) {
+      break;
+    }
+    int64_t gcd = std::gcd(srcDims[pSrc], dstDims[pDst]);
+    if (gcd == 1) {
+      return -1;
+    }
+    ++numIdxs;
+    srcDims[pSrc] /= gcd;
+    dstDims[pDst] /= gcd;
+  }
+  while (pSrc >= 0) {
+    if (srcDims[pSrc] > 1) {
+      ++numIdxs;
+    }
+    --pSrc;
+  }
+  while (pDst >= 0) {
+    if (dstDims[pDst] > 1) {
+      ++numIdxs;
+    }
+    --pDst;
+  }
+  return numIdxs;
+}
+
+Value flattenTensor(OpBuilder &builder, Value src) {
+  auto context = builder.getContext();
+  auto srcType = src.getType().cast<RankedTensorType>();
+  auto shape = srcType.getShape();
+  unsigned numIdxs = shape.size();
+  int64_t size = 1;
+
+  // Compute the Affine exprs and maps
+  SmallVector<AffineExpr, 4> srcExprs(shape.size());
+  AffineExpr dstExpr = getAffineConstantExpr(0, context);
+  for (int dim = shape.size() - 1; dim >= 0; --dim) {
+    srcExprs[dim] = getAffineDimExpr(dim, context);
+    dstExpr = dstExpr + getAffineConstantExpr(size, context) *
+                            getAffineDimExpr(dim, context);
+    size *= shape[dim];
+  }
+  AffineMap srcMap = AffineMap::get(numIdxs, 0, srcExprs, context);
+  AffineMap sinkMap = AffineMap::get(numIdxs, 0, dstExpr);
+
+  // Build contraction for the flattening
+  auto elementType = srcType.getElementType();
+  auto dstType = RankedTensorType::get({size}, elementType);
+  auto ident = tile::createIdentity(builder, builder.getUnknownLoc(),
+                                    elementType, AggregationKind::assign);
+  return builder.create<ContractionOp>(
+      builder.getUnknownLoc(),
+      /* resultType = */ dstType,
```

Comment:
format for this should be `/*resultType=*/dstType`

Reply:
done